### PR TITLE
Add the ability to get per-platform information for joypads

### PR DESCRIPTION
**Pull request by EIREXE**
This adds the ability for games to obtain platform-specific information about joypads such as their vendor/product ID, their XInput gamepad index or the real name of the device before it gets swapped out by the gamecontrollerdb's name.

The rationale behind this being returned as a `Dictionary` is that there's a big disparity between input and OS APIs, so it makes more sense to have a single "extra info" function.

This PR also includes a rebased version of #76045, this is because this PR is intended to be mainly to help implementing Steam Input glyphs, as some information needed for that is not accessible in the engine normally.

### CODE REVIEW

**Comment by EIREXE** [comment]
Updated with changes suggested by @RandomShaper on the original PR

File: platform/linuxbsd/joypad_linux.cpp
Comment (RandomShaper) [inline platform/linuxbsd/joypad_linux.cpp:424]: Maybe this string is a good candidate for a constant/macro as well?

File: platform/linuxbsd/joypad_linux.cpp
Comment (RandomShaper) [inline platform/linuxbsd/joypad_linux.cpp:425]: So this hardcoded number can be replaced by a compile-time constant.

Comment (EIREXE) [reply platform/linuxbsd/joypad_linux.cpp:425]: How would you suggest I make this compile time constant?

Comment (Calinou) [reply platform/linuxbsd/joypad_linux.cpp:425]: Probably `constexpr`, since we can use modern C++ features.

Comment (EIREXE) [reply platform/linuxbsd/joypad_linux.cpp:425]: I see, I tought he meant using constexpr to calculate the length of the string at compile time to use in substr, i'll cook up the change.

Comment (RandomShaper) [reply platform/linuxbsd/joypad_linux.cpp]: You can just add a `#define VALVE_GAMEPAD_NAME_PREFIX "Microsoft X-Box 360 pad "` and do `name.substr(strlen(VALVE_GAMEPAD_NAME_PREFIX))`. I think most compilers are able to optimize out the `strlen()` and, well, if it didn't happen for some reason, it's not the end of the world either.

Comment (EIREXE) [reply platform/linuxbsd/joypad_linux.cpp]: should be ok now!

**Review by RandomShaper — APPROVED** [review]
Looks good!

**NOTE:** This will have to wait until #76045 is merged as it may need a new rebase.

**Comment by EIREXE** [comment]
Sure, i'll wait

**Comment by EIREXE** [comment]
@RandomShaper I rebased it, not sure why the test with ASAN fails, is it an issue with CI or is there something I did wrong?

**Comment by AThousandShips** [comment]
It's a ci bug, it's being tracked :)

**Comment by RandomShaper** [comment]
Good to go then!

File: doc/classes/Input.xml
Comment (YuriSizov) [inline doc/classes/Input.xml:128]:
```suggestion
				Returns a dictionary with extra platform-specific information about the device, e.g. the raw gamepad name from the OS or the Steam Input index.
```

File: platform/linuxbsd/joypad_linux.cpp
Comment (YuriSizov) [inline platform/linuxbsd/joypad_linux.cpp:63]:
```suggestion
// On Linux with Steam Input Xbox 360 devices have an index appended to their device name, this index is
// the Steam Input gamepad index.
#define VALVE_GAMEPAD_NAME_PREFIX "Microsoft X-Box 360 pad "
// IDs used by Steam Input virtual controllers.
// See https://partner.steamgames.com/doc/features/steam_controller/steam_input_gamepad_emulation_bestpractices
```

**Comment by EIREXE** [comment]
Should be good now.

File: doc/classes/Input.xml
Comment (EIREXE) [reply doc/classes/Input.xml:128]: Added

**Comment by YuriSizov** [comment]
@EIREXE I know this seems minor, but could you please be more careful with formatting? Platform names must be correctly capitalized in your newly added documentation. You also missed one instance of that in my suggestion for "Steam Input", which is a proper name.

It's best to make sure these don't slip into your PRs to begin with, as we will point them out and it will add to your work fixing and rebasing those, and will make the PR process longer for every side. Thanks 🙏

**Comment by EIREXE** [comment]
> @EIREXE I know this seems minor, but could you please be more careful with formatting? Platform names must be correctly capitalized in your newly added documentation. You also missed one instance of that in my suggestion for "Steam Input", which is a proper name.
> 
> It's best to make sure these don't slip into your PRs to begin with, as we will point them out and it will add to your work fixing and rebasing those, and will make the PR process longer for every side. Thanks 🙏

Yeah sorry, I'm very forgetful and easily distracted, i'll take a look.

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**
